### ISSUE TITLE
[WindowsDX] Multisampling artifacts, not present in DesktopGL version of project

### ISSUE THREAD

**Issue by HopefulFrog** [MonoGame/MonoGame#7804]
So, I noticed in my project some bizarre line artifacts that would go away if multi-sampling is disabled. Here’s sample code that shows the artifact:

```c#
public class Game1 : Game
{
    GraphicsDeviceManager graphics;
    SpriteBatch spriteBatch;

    Texture2D blankTexture;
    RectangleF box;

    RenderTarget2D renderTarget;

    public Game1()
    {
        graphics = new GraphicsDeviceManager(this);
        Content.RootDirectory = "Content";
        IsFixedTimeStep = false;
        graphics.SynchronizeWithVerticalRetrace = true;
        graphics.PreferMultiSampling = true;
    }

    protected override void LoadContent()
    {
        spriteBatch = new SpriteBatch(GraphicsDevice);
        blankTexture = new Texture2D(GraphicsDevice, 1, 1);

        Color[] colors = new Color[1];
        colors[0] = new Color(255, 255, 255);

        blankTexture.SetData(colors);

        box = new RectangleF(124, 49, 100, 100);

        renderTarget = new RenderTarget2D(GraphicsDevice, graphics.PreferredBackBufferWidth, graphics.PreferredBackBufferHeight,
            false, SurfaceFormat.Color, DepthFormat.None, 8, RenderTargetUsage.PreserveContents);

        base.LoadContent();
    }

    protected override void Update(GameTime gameTime)
    {
        KeyboardState keys = Keyboard.GetState();
        if (keys.IsKeyDown(Keys.Escape))
            Exit();

        float moveSpeed = 50f;
        float moveAmount = (float)(moveSpeed * gameTime.ElapsedGameTime.TotalSeconds);

        if (keys.IsKeyDown(Keys.Up))
            MoveBox(0, -moveAmount);
        if (keys.IsKeyDown(Keys.Down))
            MoveBox(0, moveAmount);
        if (keys.IsKeyDown(Keys.Left))
            MoveBox(-moveAmount, 0);
        if (keys.IsKeyDown(Keys.Right))
            MoveBox(moveAmount, 0);

        base.Update(gameTime);

        void MoveBox(float x, float y)
        {
            box.X += x;
            box.Y += y;
        }
    }

    protected override void Draw(GameTime gameTime)
    {
        bool useRenderTarget = true;

        if (useRenderTarget)
            GraphicsDevice.SetRenderTarget(renderTarget);
        GraphicsDevice.Clear(Color.White);

        spriteBatch.Begin();
        RectangleF rectangle = box;
        DrawRectangle(spriteBatch, blankTexture, rectangle, Color.Black);
        rectangle.Inflate(-2, -2);
        DrawRectangle(spriteBatch, blankTexture, rectangle, Color.White);
        spriteBatch.End();

        if (useRenderTarget)
        {
            GraphicsDevice.SetRenderTarget(null);
            GraphicsDevice.Clear(Color.CornflowerBlue);
            spriteBatch.Begin();
            spriteBatch.Draw(renderTarget, renderTarget.Bounds, new Color(255, 255, 255));
            spriteBatch.End();
        }

        base.Draw(gameTime);
    }

    public void DrawRectangle(SpriteBatch batch, Texture2D texture, RectangleF rectangle, Color color)
    {
        if (rectangle.IsEmpty)
            return;

        Vector2 scale = rectangle.Size / texture.Bounds.Size.ToVector2();
        batch.Draw(texture, rectangle.Location, null, color, 0f, Vector2.Zero, scale, SpriteEffects.None, 1f);
    }
}

// RectangleF struct, with elements unnecessary to the example removed
public struct RectangleF
{
    public float X;
    public float Y;
    public float Width;
    public float Height;

    public Vector2 Location
    {
        get => new(X, Y);
        set
        {
            X = value.X;
            Y = value.Y;
        }
    }
    public Vector2 Size
    {
        get => new(Width, Height);
        set
        {
            Width = value.X;
            Height = value.Y;
        }
    }
    public bool IsEmpty => Width == 0 || Height == 0;

    public RectangleF(float x, float y, float width, float height)
    {
        X = x;
        Y = y;
        Width = width;
        Height = height;
    }

    public void Inflate(float xAmount, float yAmount)
    {
        X -= xAmount;
        Y -= yAmount;
        Width += xAmount * 2f;
        Height += yAmount * 2f;
    }
}
```

Here’s the output with 8 passed as the PreferredMultiSampleCount to the RenderTarget2D constructor:

![graphical artifact 1](https://user-images.githubusercontent.com/50537111/169191999-47c776b9-11bd-4545-8631-7c17fe352a8a.png)

See that little grey line in the upper-right corner of the box? That shouldn’t be there, and it goes away if you set the PreferredMultiSampleCount to 1:

![graphical artifact 2](https://user-images.githubusercontent.com/50537111/169192024-d61defd3-813d-41a8-9626-00764f6b6722.png)

It also changes and goes away if you move the box around with the arrow keys. In the sample code, it also goes away if you set PreferredMultiSampleCount to 4, but in my actual project the artifacts remain even at a PreferredMultiSampleCount of 4 (or at least, I thought it did, but I retested it and now it seems to have gone away at 4).

The artifact still shows up if you skip the rendertarget by setting useRenderTarget to false, as long as graphics.PreferMultiSampling is set to true.

If you set the RenderTargetUsage to DiscardContents, the line instead appears and disappears, flickering, if that helps.

The bug doesn’t show up at all in my OpenGL build, so it seems to be DirectX only.

Anyone have any insight as to what might be causing this bizarre behavior?

#### What version of MonoGame does the bug occur on:
- MonoGame 3.8.something (built from source relatively recently)

#### What operating system are you using:
- Windows

#### What MonoGame platform are you using:

- WindowsDX

**Comment by HopefulFrog** [comment]
Does anyone have any insight into this? I'd like to be able to include the option to enable multisampling in the DirectX build of my upcoming game, but find the visual artifacts to be unacceptable.

**Comment by HopefulFrog** [comment]
I looked through MonoGame's code, hoping that I might be able to figure out the problem myself, but it's way above my pay grade. The best I can guess is that the problem might be here, in RenderTarget2D.DirectX.cs:

```c#
internal virtual void ResolveSubresource()
{
    lock (GraphicsDevice._d3dContext)
    {
        GraphicsDevice._d3dContext.ResolveSubresource(
            GetMSTexture(),
            0,
            GetTexture(),
            0,
            SharpDXHelper.ToFormat(_format));
    }
}
```

but I wouldn't be able to tell what exactly is the source of the problem, or even if it's a MonoGame bug at all. It might be a SharpDX bug for all I know.

So, for the Windows version of my game I have to decide between having decreased performance (from using the DesktopGL build) or not including the option to enable multi-sampling (because of the visual artifacts in the DirectX build). Unless there's a way for me to roll my own shader to do the multisampling...

**Comment by EzeKees** [comment]
It looks like a culling error. Test if this artifact disappears when editing the rasterizerState values

**Comment by KakCAT** [comment]
Any chances this is a driver problem? I've run the code and it seems to work well to me with DX and multisampling, although I must admit I'm not running the latest MG version (I'm running a 6 months old build)

**Comment by HopefulFrog** [comment]
> It looks like a culling error. Test if this artifact disappears when editing the rasterizerState values

I tested passing the different RasterizerState values to spriteBatch.Begin both before the render target is drawn to and the render target itself is drawn. CullNone and CullCounterClockwise both still exhibited the problem, while CullClockwise being passed to either Begin method simply resulted in a solid color blank screen.


> Any chances this is a driver problem? I've run the code and it seems to work well to me with DX and multisampling, although I must admit I'm not running the latest MG version (I'm running a 6 months old build)

Huh, weird. I suppose it could be related to drivers in some obscure way. I do think my MonoGame build is more recent than that. The issue might very well be with my SharpDX DLLs. I don't remember how I got them, but they might not be up to date.

In the meantime, I found a workaround for the problem. Rotating the rectangle by an imperceptible amount (0.001f) causes the line artifact to disappear. It seems it only shows up when the two rectangles are _exactly_ lined up.

**Comment by nkast** [comment]
Can you tey to enable depth on the rendertarget?

**Comment by HopefulFrog** [comment]
> Can you try to enable depth on the rendertarget?
> 
> Also what Card do you have?
> Do you have the latest drivers from the vendor?
> Have you tested this on other systems?

I tried different values besides None for the DepthFormat passed to the render target constructor, but that didn't seem to change anything. However, I tried messing around with the SurfaceFormat and discovered that with the SurfaceFormat set to Rgba64 or Vector4, the problem disappears. No artifact. I assumed that SurfaceFormat.Color was the closest thing there was to a "default" value, but is that not the case? I don't know enough to know why the issue would present itself with one SurfaceFormat but not another.

I'm using a 1660 Super. I don't think I have the latest drivers per se. They might be half a year to a year old. I haven't tested the problem on other systems yet.

If there's no downside to using a different SurfaceFormat, then that's a good enough solution for my purposes.
